### ISSUE TITLE
Walking the infinite table rolling over too far.

### ISSUE THREAD

**Issue by johadalin** [Metaswitch/chronos#188]
Calling snmpwalk over the infinite table, by means of one of the proxied OIDs, (e.g. snmpwalk -v2c -c clearwater localhost 1.2.826.0.1.1578918.9.10.6) rolls over beyond the end of the tag sent in and continues walking apparently infinitely until constrained by memory limits. 

an excerpt from a log showing this issue is as follows:
```
02-11-2015 16:44:48.412 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag CAL^V, identifier <5, 2>
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.4.67.65.76.127.2.1
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <2, 2>
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.4.67.65.76.127.2.2 with tag CAL^? and identifier <2, 2>
02-11-2015 16:44:48.412 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag CAL^?, identifier <2, 2>
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.4.67.65.76.107.3.3
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <4, 1>
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.4.67.65.76.107.4.1 with tag CALk and identifier <4, 1>
02-11-2015 16:44:48.412 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag CALk, identifier <4, 1>
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.4.67.65.76.212.5.3
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:183: This request goes beyond the table
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.4.67.65.76.158.2.2
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <2, 3>
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.4.67.65.76.158.2.3 with tag CAL<9E> and identifier <2, 3>
02-11-2015 16:44:48.412 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag CAL<9E>, identifier <2, 3>
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.4.67.65.76.235.3.2
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <3, 3>
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.4.67.65.76.235.3.3 with tag CAL<EB> and identifier <3, 3>
02-11-2015 16:44:48.412 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag CAL<EB>, identifier <3, 3>
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.4.67.65.76.22.5.2
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <5, 3>
02-11-2015 16:44:48.412 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.4.67.65.76.22.5.3 with tag CAL^V and identifier <5, 3>
02-11-2015 16:44:48.412 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
```

after reading CALL, the walk continues to CALM etc. when instead it should simply stop the walk.

**Comment by johadalin** [comment]
this is probably related to issue #185

**Comment by johadalin** [comment]
Below is another log excerpt from the chronos logs showing the point at which the OID walk corrupts:
```
03-11-2015 09:48:28.039 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
03-11-2015 09:48:28.040 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.4.67.65.76.76.5.1
03-11-2015 09:48:28.040 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <5, 2>
03-11-2015 09:48:28.040 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.4.67.65.76.76.5.2 with tag CALL and identifier <5, 2>
03-11-2015 09:48:28.040 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
03-11-2015 09:48:28.040 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag CALL, identifier <5, 2>
03-11-2015 09:48:28.040 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
03-11-2015 09:48:28.041 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
03-11-2015 09:48:28.041 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.4.67.65.76.76.5.2
03-11-2015 09:48:28.041 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <5, 3>
03-11-2015 09:48:28.041 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.4.67.65.76.76.5.3 with tag CALL and identifier <5, 3>
03-11-2015 09:48:28.042 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
03-11-2015 09:48:28.042 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag CALL, identifier <5, 3>
03-11-2015 09:48:28.042 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
03-11-2015 09:48:28.043 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
03-11-2015 09:48:28.043 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.4.67.65.76.76.5.3
03-11-2015 09:48:28.043 UTC Debug snmp_infinite_timer_count_table.cpp:183: This request goes beyond the table
03-11-2015 09:48:28.043 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
03-11-2015 09:48:28.043 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999
03-11-2015 09:48:28.044 UTC Debug snmp_infinite_timer_count_table.cpp:177: Invalid GET request
03-11-2015 09:48:28.044 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
03-11-2015 09:48:28.044 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999
03-11-2015 09:48:28.044 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <2, 1>
03-11-2015 09:48:28.044 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.65.2 with tag  and identifier <2, 1>
03-11-2015 09:48:28.044 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
03-11-2015 09:48:28.044 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag , identifier <2, 1>
03-11-2015 09:48:28.044 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
03-11-2015 09:48:28.044 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
03-11-2015 09:48:28.044 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.65.2
03-11-2015 09:48:28.045 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <2, 1>
03-11-2015 09:48:28.045 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.65.2.65.2 with tag ▒4▒▒~ and identifier <2, 1>
03-11-2015 09:48:28.045 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
03-11-2015 09:48:28.045 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag ▒4▒▒~, identifier <2, 1>
03-11-2015 09:48:28.045 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
03-11-2015 09:48:28.045 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
03-11-2015 09:48:28.045 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.65.2.65.2
03-11-2015 09:48:28.045 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <2, 1>
03-11-2015 09:48:28.045 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.65.2.65.2.1.2 with tag A▒▒~ and identifier <2, 1>
03-11-2015 09:48:28.045 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
03-11-2015 09:48:28.045 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag A▒▒~, identifier <2, 1>
03-11-2015 09:48:28.045 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
03-11-2015 09:48:28.046 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
03-11-2015 09:48:28.046 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.65.2.65.2.1.2
03-11-2015 09:48:28.046 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <2, 1>
03-11-2015 09:48:28.046 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.65.2.65.2.1.2.139632876453920.2 with tag A~ and identifier <2, 1>
03-11-2015 09:48:28.046 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
03-11-2015 09:48:28.046 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag A~, identifier <2, 1>
03-11-2015 09:48:28.046 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
03-11-2015 09:48:28.046 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
03-11-2015 09:48:28.046 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.65.2.65.2.1.2.3489660960.2
03-11-2015 09:48:28.046 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <2, 1>
03-11-2015 09:48:28.046 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.65.2.65.2.1.2.3489660960.2.2.1 with tag A  and identifier <2, 1>
03-11-2015 09:48:28.047 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
03-11-2015 09:48:28.047 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag A , identifier <2, 1>
03-11-2015 09:48:28.047 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
03-11-2015 09:48:28.047 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
03-11-2015 09:48:28.047 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.65.2.65.2.1.2.3489660960.2.2.1
03-11-2015 09:48:28.047 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <2, 1>
03-11-2015 09:48:28.047 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID iso.2.826.0.1.1578918.999.65.2.65.2.1.2.3489660960.2.2.1.0.0 with tag A ▒E▒▒~ and identifier <2, 1>
03-11-2015 09:48:28.047 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
03-11-2015 09:48:28.047 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag A ▒E▒▒~, identifier <2, 1>
03-11-2015 09:48:28.047 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests
03-11-2015 09:48:28.047 UTC Debug snmp_infinite_timer_count_table.cpp:121: Starting handling batch of SNMP requests
03-11-2015 09:48:28.048 UTC Debug snmp_infinite_timer_count_table.cpp:130: Handling SNMP request for OID iso.2.826.0.1.1578918.999.65.2.65.2.1.2.3489660960.2.2.1.0.0
03-11-2015 09:48:28.048 UTC Debug snmp_infinite_timer_count_table.cpp:355: Identifier updated to valid form: <2, 1>
03-11-2015 09:48:28.048 UTC Debug snmp_infinite_timer_count_table.cpp:208: Parsed SNMP request to OID  with tag A  and identifier <2, 1>
03-11-2015 09:48:28.048 UTC Debug timer_counter.cpp:108: Bringing the statistics up to date
03-11-2015 09:48:28.048 UTC Debug snmp_infinite_timer_count_table.cpp:216: Got value 0 for tag A , identifier <2, 1>
03-11-2015 09:48:28.048 UTC Debug snmp_infinite_timer_count_table.cpp:228: Finished handling batch of SNMP requests

**Comment by bossmc** [comment]
Having fixed #185, walking a proxied OID gives

```
$ snmpwalk -c public -v2c localhost $index
iso.2.826.0.1.1578918.9.10.7.1.2.1 = Gauge32: 0
iso.2.826.0.1.1578918.9.10.7.1.2.2 = Gauge32: 0
iso.2.826.0.1.1578918.9.10.7.1.2.3 = Gauge32: 0
iso.2.826.0.1.1578918.9.10.7.1.3.1 = Gauge32: 0
iso.2.826.0.1.1578918.9.10.7.1.3.2 = Gauge32: 0
iso.2.826.0.1.1578918.9.10.7.1.3.3 = Gauge32: 0
iso.2.826.0.1.1578918.9.10.7.1.4.1 = Gauge32: 0
iso.2.826.0.1.1578918.9.10.7.1.4.2 = Gauge32: 0
iso.2.826.0.1.1578918.9.10.7.1.4.3 = Gauge32: 0
iso.2.826.0.1.1578918.9.10.7.1.5.1 = Gauge32: 0
iso.2.826.0.1.1578918.9.10.7.1.5.2 = Gauge32: 0
iso.2.826.0.1.1578918.9.10.7.1.5.3 = Gauge32: 0
```

...and stops there.
